extensions/catalog/federated-catalog-cache/src/main/java/org/eclipse/dataspaceconnector/catalog/store/InMemoryFederatedCacheStore.java

```diff
@@ -32,7 +32,7 @@
  */
 public class InMemoryFederatedCacheStore implements FederatedCacheStore {
 
-    private final Map<String, ContractOffer> cache = new ConcurrentHashMap<>();
+    private final Map<String, MarkableEntry<ContractOffer>> cache = new ConcurrentHashMap<>();
```

Comment:
Several remarks:

1/ Is there really a need for a generic in the MarkableEntry? 

Also I would prefer to keep it immutable:

~~~
private static class MarkableEntry {
  private final ContractOffer contractOffer;
  private final isPotentiallyExpired;

  public MarkableEntry(boolean isPotentiallyExpired, ContractOffer contractOffer) {
    this.contractOffer = contractOffer;
    this.isPotentiallyExpired= isPotentiallyExpired;
  }
}
~~~

2/ Then in the `save` function becomes:

~~~
public void save(ContractOffer contractOffer) {
        lockManager.writeLock(() -> cache.put(contractOffer.getAsset().getId(), new MarkableEntry(false, contractOffer)));
    }
~~~

3/ And finally instead of having two methods for marking, and then deleting the deprecated, I would merge them in a single one:

~~~
@Override
public void deleteExpired() {
    lockManager.writeLock({
       cache.values().removeIf(me::isExpired);
       cache.replaceAll((k,v) -> v = new MarkableEntry(true, v.getContractOffer()));
       return null;
    }
}
~~~

Reply:
i'm fine with the first two points. I guess the in-mem impl is intended for development/testing purposes only, so I won't go into whether modifying or replacing a map entry is better, faster, more efficient, etc.

The reason why I kept the `markAll` and `removeMarked` methods separate is that the behaviour for different store implementations than be controlled by the owning object (here: the `LoaderImpl`); it makes it apparent through the interface that these are two distinct steps.

If we collate them into one method call, it would push business logic down into the store implementation, which I find dangerous.